gcsio/src/main/java/com/google/cloud/hadoop/gcsio/MetricsPublisher.java

```diff
@@ -0,0 +1,240 @@
+import io.opencensus.stats.StatsRecorder;
+import io.opencensus.stats.View;
+import io.opencensus.stats.View.Name;
+import io.opencensus.stats.ViewManager;
+import io.opencensus.tags.TagContext;
+import io.opencensus.tags.TagContextBuilder;
+import io.opencensus.tags.TagKey;
+import io.opencensus.tags.TagMetadata;
+import io.opencensus.tags.TagMetadata.TagTtl;
+import io.opencensus.tags.TagValue;
+import io.opencensus.tags.Tagger;
+import io.opencensus.tags.Tags;
+import java.io.IOException;
+import java.util.List;
+import javax.annotation.concurrent.GuardedBy;
+
+/** Interface for exposing utility methods to publish metrics via open-census api-spec. */
+interface MetricsPublisher {
+
+  /**
+   * Publishes metric for specified open-census measurement
+   *
+   * @param key represents filterable attribute
+   * @param value for the TagKey
+   * @param ml - Measurement to be recorded
+   * @param n - Value of the measurement recorded
+   */
+  void recordTaggedStat(TagKey key, String value, MeasureLong ml, Long n);
+
+  /**
+   * Publishes metric for specified open-census measurement
+   *
+   * @param keys - List of TagKeys attributes used for filtration
+   * @param values - corresponding values for list of TagKeys
+   * @param ml - Measurement to be recorded
+   * @param n - Value of the measurement recorded
+   */
+  void recordLong(TagKey[] keys, String[] values, MeasureLong ml, Long n);
+}
+
+/** Publishes open-census measurements to StackDriver */
+class CloudMonitoringMetricsPublisher implements MetricsPublisher {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting static final Object monitor = new Object();
+
+  @GuardedBy("monitor")
+  private static boolean initialized = false;
+
+  private static final TagMetadata TAG_METADATA_NO_PROPAGATION =
+      TagMetadata.create(TagTtl.NO_PROPAGATION);
+
+  static final MeasureLong LATENCY_MS =
+      MeasureLong.create("gcsio/latency", "The latency in milliseconds ", "ms");
+
+  static final MeasureLong MESSAGE_LATENCY_MS =
+      MeasureLong.create(
+          "gcsio/message/latency", "The latency in milliseconds per gcs message loop", "ms");
+
+  static final MeasureLong REQUESTS =
+      MeasureLong.create(
+          "gcsio/requests", "The distribution of retry attempts for gcs calls", "By");
+
+  static final MeasureLong REQUEST_RETRIES =
+      MeasureLong.create("gcsio/retries", "The distribution of retry attempts for gcs calls", "By");
+
+  static final TagKey METHOD = TagKey.create("method");
+  static final TagKey STATUS = TagKey.create("status");
+  static final TagKey ERROR = TagKey.create("error");
+  static final TagKey THREAD = TagKey.create("thread");
+  static final TagKey PROTOCOL = TagKey.create("protocol");
+
+  static final List<TagKey> TAG_KEYS = ImmutableList.of(METHOD, STATUS, ERROR, THREAD, PROTOCOL);
+
+  static final List<Double> RPC_MILLIS_BUCKET_BOUNDARIES =
+      ImmutableList.of(
```

Comment:
It is based on OpenCensus lib impl for grpc, so that we align on buckets for both standard and custom metrics. Unfortunately it isn't public for us to use directly